src/economy/economy.cpp

```diff
@@ -519,64 +546,120 @@ float artisan_scale_limit(sys::state& state, dcon::nation_id n, dcon::commodity_
 	return least;
 }
 
-void adjust_artisan_balance(sys::state& state, dcon::nation_id n) {
-	auto const csize = state.world.commodity_size();
-	float distribution_drift_speed = 0.01f;
+bool valid_artisan_good(sys::state& state, dcon::nation_id n, dcon::commodity_id cid) {
+	auto kf = state.world.commodity_get_key_factory(cid);
+	return (
+		state.world.commodity_get_artisan_output_amount(cid) > 0.0f
+		&& (
+			state.world.commodity_get_is_available_from_start(cid)
+			|| (
+				kf && state.world.nation_get_active_building(n, kf)
+				)
+			)
+		);
+}
 
-	auto num_artisans = state.world.nation_get_demographics(n, demographics::to_key(state, state.culture_definitions.artisans));
+inline constexpr float ln_2 = 0.30103f;
 
-	std::vector<float> profits;
-	profits.resize(csize);
+//crude approximation of exp
+float pseudo_exp_for_negative(float f) {
+	if(f < -128.f) {
+		return 0.f;
+	}
 
-	{
-		float total_weights = 0.0f;
-		float total_ex_weights = 0.0f;
-		float total_possible = 0.0f;
+	f = f / 128.f;
+	f = 1 + f + f * f / 2 + f * f * f / 6;
 
-		for(uint32_t i = 1; i < csize; ++i) {
-			dcon::commodity_id cid{ dcon::commodity_id::value_base_t(i) };
-			auto kf = state.world.commodity_get_key_factory(cid);
+	f = f * f; // 2
+	f = f * f; // 4
+	f = f * f; // 8
+	f = f * f; // 16
+	f = f * f; // 32
+	f = f * f; // 64
+	f = f * f; // 128
 
-			if(state.world.commodity_get_artisan_output_amount(cid) > 0.0f && (state.world.commodity_get_is_available_from_start(cid) || (kf && state.world.nation_get_active_building(n, kf)))) {
-				//auto limit = artisan_scale_limit(state, n, cid);
-				auto profit = base_artisan_profit(state, n, cid); //* limit;
-				profits[cid.index()] = profit;
+	return f;
+}
 
-				if(profit > 0.f) {
-					total_weights += profit;
-					total_possible += 1.0f;
-				}
-			}
+float get_artisans_multiplier(sys::state& state, dcon::nation_id n) {
+	float multiplier = 0.000001f * state.world.nation_get_everyday_needs_costs(n, state.culture_definitions.artisans);
+	return 1.f / (multiplier + 1.f);
+}
+
+void scores_to_distribution(std::vector<float> & scores, std::vector<float> & distribution, float multiplier) {
+	float max_score = std::numeric_limits<float>::min();
```

Comment:
I'll quote from the reference here
> For floating-point types with denormalization, min() returns the minimum positive normalized value. Note that this behavior may be unexpected, especially when compared to the behavior of min() for integral types. To find the value that has no values less than it, use [lowest()](https://en.cppreference.com/w/cpp/types/numeric_limits/lowest).
You probably meant 0? or negative infinity?

Reply:
I meant negative infinity, thanks. But as one of values was guaranteed to be positive, it doesn't influence anything. Will change to lowest.